postgrest/base_request_builder.py

```diff
@@ -405,3 +413,10 @@ def single(self: _FilterT) -> _FilterT:
         """
         self.headers["Accept"] = "application/vnd.pgrst.object+json"
         return self
+
+    def maybe_single(self: _FilterT) -> _FilterT:
```

Comment:
So I was initially imagining that we'd return a MaybeSingleRequestBuilder here instead of returning `self` like in the other methods; this eliminates the need for the new SyncQueryBuilder class

Reply:
`execute` method is from `SyncQueryBuilder` which later was sourced by `SyncSelectRequestBuilder`. This class would also source filters method (`single`, `order`, etc) from `BaseSelectRequestBuilder`. 

In this case, class-diagram wise, I'm not really sure if we return `MaybeSingleRequestBuilder` from `maybe_single` in `SyncSelectRequestBuilder` would override the `execute` method from other class (in this case, `SyncQueryBuilder`). 

Another reason why I go for this design is for a better code readability. I think the design you mentioned would incur longer time to understand as the class design gets more complicated. Plus, using this design (adding custom headers) would follow all the other filter method, namely returning `self` and adding attributes to headers/params. 

What do you think? @anand2312

Reply:
While I agree the class design gets more complicated, the current implementation cannot be statically type-checked. (whether single/maybe_single was used can only be known at runtime)
> In this case, class-diagram wise, I'm not really sure if we return MaybeSingleRequestBuilder from maybe_single in SyncSelectRequestBuilder would override the execute method from other class (in this case, SyncQueryBuilder).
MaybeSingleRequestBuilder would implement `execute`, so returning an instance of it would ensure that _that_ execute method is used instead of the usual one.
> Plus, using this design (adding custom headers) would follow all the other filter method, namely returning self and adding attributes to headers/params.
I actually think that enforcing `single` and `maybe_single` be the last methods to be called before execute isn't a bad idea. @leynier also had proposed this (see #102 ). Imo calling more filters after single/maybe_single just looks kinda odd.

I want this to be statically type-checkable because I have plans to properly type `APIResponse` (and create a separate SingleAPIResponse). Right now, APIResponse.data is typed as `Any`, which just turns the type checker off, which isn't really good UX for end-users.

Reply:
Hi @anand2312, if we decide to go for this method, then we won't be able to utilize inheritance in the classes since it means that the parent's class return type must be the same as the children's. In our implementation, we want `execute` to have `APIResponse` type return and a new class for accessing new `execute` that returns `SingleAPIResponse` for `single` or `maybe_single` method. So, should we proceed on doing this? If so, there'll be quite some code duplication. If it's a go, then I shall implement it

Reply:
@bariqhibat  Yeah, I want to avoid the Union.
In my opinion we don't need inheritance. The MaybSingleRequestBuilder / SingleRequestBuilder only needs the `execute` method and nothing else (no filters need to be allowed after a call to `single`/`maybe_single`). I can't think of any other code being duplicated, or am I missing something?

Reply:
They'd need the constructor as well. Anyway, I'll pushthe new changes by tomorrow